trieste/acquisition/function/function.py

```diff
@@ -223,12 +232,132 @@ def __call__(self, x: TensorType) -> TensorType:
+        :param eta: The "best" observation.
+        :return: The expected improvement function. This function will raise
+            :exc:`ValueError` or :exc:`~tf.errors.InvalidArgumentError` if used with a batch size
+            greater than one.
+        """
+        self._model = model
+        self._eta = tf.Variable(eta)
+
+    def update(self, eta: TensorType) -> None:
+        """Update the acquisition function with a new eta value."""
+        self._eta.assign(eta)
+
+    @tf.function
+    @check_shapes(
+        "x: [N..., 1, D] # This acquisition function only supports batch sizes of one",
+        "return: [N..., L]",
+    )
+    def __call__(self, x: TensorType) -> TensorType:
+        eps = (
+            tf.constant(-1e12, dtype=tf.float64)
+            if x.dtype == tf.float64
+            else tf.constant(-1e6, dtype=tf.float32)
+        )
+        mean, variance = self._model.predict(tf.squeeze(x, -2))
+        sigma = tf.sqrt(tf.clip_by_value(variance, clip_value_min=eps, clip_value_max=x.dtype.max))
+        u = (self._eta - mean) / sigma
+
+        return _log_ei_helper(u) + tf.math.log(sigma)
+
+
+def _log_ei_helper(u: TensorType) -> TensorType:
+    """Accurately computes log(phi(u) + u * Phi(u)) in a differentiable manner for u in
+    [-10^100, 10^100] in double precision, and [-10^20, 10^20] in single precision.
+    The implementation is inspired by the BoTorch implementation
+    (https://github.com/pytorch/botorch/blob/main/botorch/acquisition/analytic.py)
+    """
+    if not (u.dtype == tf.float32 or u.dtype == tf.float64):
+        raise TypeError(
+            f"LogExpectedImprovement only supports torch.float32 and torch.float64 "
+            f"dtypes, but received {u.dtype=}."
+        )
+    # The function has two branching decisions. The first is u < bound, and in this
+    # case, just taking the logarithm of the naive _ei_helper implementation works.
+
+    bound = tf.constant(-1.0, dtype=u.dtype)
+    u_upper = tf.where(u < bound, bound, u)  # mask u to avoid NaNs in gradients
+    normal = tfp.distributions.Normal(
+        loc=tf.constant(0.0, dtype=u.dtype), scale=tf.constant(1.0, dtype=u.dtype)
+    )
+
+    log_ei_upper = tf.math.log(normal.prob(u_upper) + u_upper * normal.cdf(u_upper))
+
+    # When u <= bound, we need to be more careful and rearrange the EI formula as
+    # log(phi(u)) + log(1 - exp(w)), where w = log(abs(u) * Phi(u) / phi(u)).
+    # To this end, a second branch is necessary, depending on whether or not u is
+    # smaller than approximately the negative inverse square root of the machine
+    # precision. Below this point, numerical issues in computing log(1 - exp(w)) occur
+    # as w approaches zero from below, even though the relative contribution to log_ei
+    # vanishes in machine precision at that point.
+    neg_inv_sqrt_eps = (
+        tf.constant(-1e6, dtype=tf.float64)
+        if u.dtype == tf.float64
+        else tf.constant(-1e3, dtype=tf.float32)
+    )
```

Comment:
these constants seem a bit too large?

Reply:
Agreed, but it is consistent with original implementation: https://github.com/pytorch/botorch/blob/01aa45e2c1287b53b9ce15f683e27be8a4ad78c5/botorch/acquisition/analytic.py#L1000

Reply:
ok I guess